src/mixins/themeable.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * A representation of the css-module class names
+ * to be applied where each class in appliedClasses
+ * is used.
+ */
+export type CSSModuleClassNames = {
+	[key: string]: boolean;
+}
+
+/**
+ * The object returned by getClasses.
+ */
+export type AppliedClasses<T> = {
+	[P in keyof T]?: CSSModuleClassNames;
+};
+
+type StringIndexedObject = { [key: string]: string; };
+
+/**
+ * Properties required for the themeable mixin
+ */
+export interface ThemeableProperties {
+	theme?: {};
+	overrideClasses?: {};
+}
+
+/**
+ * Themeable Options
+ */
+export interface ThemeableOptions {
+	properties: ThemeableProperties;
+}
+
+/**
+ * Themeable Mixin
+ */
+export interface ThemeableMixin<P> extends Evented {
+	theme: AppliedClasses<P>;
+}
+
+/**
+ * Themeable
+ */
+export interface Themeable<P> extends ThemeableMixin<P> {
+	baseTheme: P;
+	properties: ThemeableProperties;
+}
+
+/**
+ * Compose Themeable Factory interface
+ */
+export interface ThemeableFactory extends ComposeFactory<ThemeableMixin<{}>, ThemeableOptions> {}
+
+/**
+ * Private map for the widgets themeClasses.
+ */
+const themeClassesMap = new WeakMap<ThemeableMixin<{}>, AppliedClasses<any>>();
+
+function addClassNameToCSSModuleClassNames(cssModuleClassNames: CSSModuleClassNames, classList: StringIndexedObject, className: string) {
+	if (classList.hasOwnProperty(className)) {
+		// split out the classname because css-module composition combines class names with a space
+		const generatedClassNames: string[] = classList[className].split(' ');
+		generatedClassNames.forEach((generatedClassName) => {
+			cssModuleClassNames[generatedClassName] = true;
+		});
+	}
+}
+
+function negatePreviousClasses<T>(previousClasses: AppliedClasses<T>, newClasses: AppliedClasses<T>) {
+	return Object.keys(previousClasses).reduce((newAppliedClasses, className: keyof T) => {
+		const oldCSSModuleClassNames: CSSModuleClassNames = previousClasses[className] || {};
+
+		const negatedCSSModuleClassNames = Object.keys(oldCSSModuleClassNames).reduce((newCSSModuleClassNames, oldCSSModuleClassName) => {
+			const currentClassNameFlag = oldCSSModuleClassNames[<keyof T> oldCSSModuleClassName];
+			// If it's true it needs to be negated and passed along, If it's false,
+			// don't return it as maquette will already have removed it.
+			if (currentClassNameFlag) {
+				newCSSModuleClassNames[oldCSSModuleClassName] = false;
+			}
+			return newCSSModuleClassNames;
+		}, <CSSModuleClassNames> {});
+
+		const calculatedClassNameMap = assign({}, negatedCSSModuleClassNames, newClasses[className]);
+		newAppliedClasses[className] = calculatedClassNameMap;
+
+		return newAppliedClasses;
+	}, <AppliedClasses<T>> {});
+}
+
+function generateThemeClasses<I, T>(instance: Themeable<I>, baseTheme: T, theme: {} = {}, overrideClasses: {} = {}) {
+	return Object.keys(baseTheme).reduce((newAppliedClasses, className: keyof T) => {
+		const newCSSModuleClassNames: CSSModuleClassNames = {};
+
+		let themeClassSource: {} = baseTheme;
```

Comment:
Thought about that but I thought we preferred if/ else over ternary for readability.